From 8f192f40316fdecf68a2863b1beb36635c9fc6fa Mon Sep 17 00:00:00 2001
From: Pavel Zwerschke <pavel.zwerschke@quantco.com>
Date: Fri, 9 Oct 2026 09:08:13 +0200
Subject: [PATCH] Build with conda-forge toolchain and system glfw/freetype

- Run a generic CMake/Ninja build from setup.py that honors CMAKE_ARGS
  instead of hardcoded per-platform commands (Visual Studio paths, etc.).
- Use glfw (and freetype on Windows) from the host environment instead of
  the vendored git submodules.
- Don't bundle vcruntime140_1.dll; conda provides the VC runtime.
- Include <mutex> in mvContext.h for std::recursive_mutex; newer libc++
  no longer provides it transitively.
---
 setup.py                  | 54 ++++++++++++---------------------------
 src/distribution.cmake    |  3 +--
 src/mvContext.h           |  1 +
 thirdparty/CMakeLists.txt | 12 ++-------
 4 files changed, 20 insertions(+), 50 deletions(-)

diff --git a/setup.py b/setup.py
index be68c1c..c13a244 100644
--- a/setup.py
+++ b/setup.py
@@ -8,6 +8,7 @@
 import sys
 import shutil
 import subprocess
+import shlex
 
 wip_version = "2.3"
 
@@ -64,41 +65,20 @@ def run(self):
         self.announce('Using readthedocs hack',level=distutils.log.INFO)
         return
 
-    if get_platform() == "Windows":
-        command = [r'set PATH="C:\Program Files (x86)\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin";"C:\Program Files (x86)\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin";"C:\Program Files (x86)\Microsoft Visual Studio\2022\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin";%PATH% && ']
-        command.append("mkdir cmake-build-local && ")
-        command.append("cd cmake-build-local && ")
-        command.append('cmake .. -G "Visual Studio 17 2022" -A "x64" -DMVDIST_ONLY=True -DMVDPG_VERSION=')
-        command.append(version_number() + " -DMV_PY_VERSION=")
-        command.append(str(sys.version_info[0]) + "." + str(sys.version_info[1]) + " && ")
-        command.append("cd .. && cmake --build cmake-build-local --config Release")
-        self.announce('Running command: %s' % "Dear PyGui Build for Windows", level=distutils.log.INFO)
-        subprocess.check_call(''.join(command), env=os.environ, shell=True)
-        src_path = os.path.dirname(os.path.abspath(__file__))
-        shutil.copy("cmake-build-local/DearPyGui/Release/_dearpygui.pyd", src_path +"/output/dearpygui")
-
-    elif get_platform() == "Linux":
-        command = ["mkdir cmake-build-local; "]
-        command.append("cd cmake-build-local; ")
-        command.append('cmake .. -DMVDIST_ONLY=True -DMVDPG_VERSION='+version_number()+ " -DMV_PY_VERSION="+ str(sys.version_info[0]) + "." + str(sys.version_info[1])+"; ")
-        command.append("cd ..; cmake --build cmake-build-local --config Release")
-        self.announce('Running command: %s' % "Dear PyGui Build for Linux",level=distutils.log.INFO)
-        subprocess.check_call(''.join(command), shell=True)
-        src_path = os.path.dirname(os.path.abspath(__file__))
-        shutil.copy("cmake-build-local/DearPyGui/_dearpygui.so", src_path +"/output/dearpygui")
-    
-    elif get_platform() == "OS X":
-        command = ["mkdir cmake-build-local; "]
-        command.append("cd cmake-build-local; ")
-        command.append('cmake .. -DMVDIST_ONLY=True -DMVDPG_VERSION='+version_number()+ " -DMV_PY_VERSION="+ str(sys.version_info[0]) + "." + str(sys.version_info[1])+"; ")
-        command.append("cd ..; cmake --build cmake-build-local --config Release")
-        self.announce('Running command: %s' % "Dear PyGui Build for OS X",level=distutils.log.INFO)
-        subprocess.check_call(''.join(command), shell=True)
-        src_path = os.path.dirname(os.path.abspath(__file__))
-        shutil.copy("cmake-build-local/DearPyGui/_dearpygui.so", src_path +"/output/dearpygui")
-
-    else:
-        self.announce('Command not ready.',level=distutils.log.INFO)
+    build_dir = "cmake-build-local"
+    cmake_args = shlex.split(os.environ.get("CMAKE_ARGS", ""), posix=get_platform() != "Windows")
+    subprocess.check_call(
+        ["cmake", "-S", ".", "-B", build_dir, "-G", "Ninja",
+         "-DCMAKE_BUILD_TYPE=Release",
+         "-DMVDIST_ONLY=True",
+         "-DMVDPG_VERSION=" + version_number(),
+         "-DMV_PY_VERSION=" + str(sys.version_info[0]) + "." + str(sys.version_info[1]),
+         "-DPython_EXECUTABLE=" + sys.executable,
+        ] + cmake_args)
+    subprocess.check_call(["cmake", "--build", build_dir, "--config", "Release"])
+    src_path = os.path.dirname(os.path.abspath(__file__))
+    ext = "_dearpygui.pyd" if get_platform() == "Windows" else "_dearpygui.so"
+    shutil.copy(os.path.join(build_dir, "DearPyGui", ext), src_path + "/output/dearpygui")
 
 class BuildPyCommand(build_py.build_py):
   def run(self):
@@ -157,8 +137,6 @@ def setup_package():
 
         # copy add items to temporary location
         shutil.copy(src_path + "/dearpygui/_dearpygui.pyi", src_path + "/output/dearpygui")
-        if get_platform() == "Windows":
-            shutil.copy(src_path + "/thirdparty/Microsoft/vcruntime140_1.dll", src_path + "/output/dearpygui")
 
     metadata = dict(
         name='dearpygui',                                      # Required
@@ -206,7 +184,7 @@ def setup_package():
     if os.environ.get('READTHEDOCS') == 'True':
         metadata['package_data']['dearpygui'] = ["__init__.py", "_dearpygui.py", "dearpygui.py", "demo.py", "experimental.py"]
     elif get_platform() == "Windows":
-        metadata['package_data']['dearpygui'] = ["__init__.py", "_dearpygui.so", "_dearpygui.pyd", "_dearpygui.pyi", "dearpygui.py", "demo.py", "experimental.py", "type_info.py", "vcruntime140_1.dll"]
+        metadata['package_data']['dearpygui'] = ["__init__.py", "_dearpygui.so", "_dearpygui.pyd", "_dearpygui.pyi", "dearpygui.py", "demo.py", "experimental.py", "type_info.py"]
     else:
         metadata['package_data']['dearpygui'] = ["__init__.py", "_dearpygui.so", "_dearpygui.pyd", "_dearpygui.pyi", "dearpygui.py", "demo.py", "experimental.py", "type_info.py"]
 
diff --git a/src/distribution.cmake b/src/distribution.cmake
index c943053..a8b5d9a 100644
--- a/src/distribution.cmake
+++ b/src/distribution.cmake
@@ -25,7 +25,6 @@ target_include_directories(_dearpygui
 
 target_link_directories(_dearpygui 
 		PRIVATE 
-			"$<$<PLATFORM_ID:Linux>:/usr/lib>"
 			${Python_LIBRARY_DIRS}
 	)
 
@@ -45,7 +44,7 @@ if(WIN32)
 
 	set_target_properties(_dearpygui PROPERTIES SUFFIX ".pyd")
 	set_target_properties(_dearpygui PROPERTIES PREFIX "")
-	target_link_libraries(_dearpygui PUBLIC d3d11 dxgi dwmapi ${Python_LIBRARIES} freetype)
+	target_link_libraries(_dearpygui PUBLIC d3d11 dxgi dwmapi ${Python_LIBRARIES} Freetype::Freetype)
 
 elseif(APPLE)
 
diff --git a/src/mvContext.h b/src/mvContext.h
index 14333b6..3bcc0f1 100644
--- a/src/mvContext.h
+++ b/src/mvContext.h
@@ -5,6 +5,7 @@
 #include <string>
 #include <future>
 #include <atomic>
+#include <mutex>
 #include <memory>
 #include "mvCore.h"
 #include "mvPyUtils.h"
diff --git a/thirdparty/CMakeLists.txt b/thirdparty/CMakeLists.txt
index 4e4f721..a437582 100644
--- a/thirdparty/CMakeLists.txt
+++ b/thirdparty/CMakeLists.txt
@@ -1,13 +1,5 @@
-﻿
-set(GLFW_BUILD_DOCS OFF)
-set(GLFW_BUILD_EXAMPLES OFF)
-set(GLFW_BUILD_TESTS OFF)
-set(GLFW_INSTALL OFF)
 
+find_package(glfw3 REQUIRED)
 if(WIN32)
-    add_subdirectory("freetype")
-elseif(APPLE)
-    add_subdirectory ("glfw")
-else() #Linux
-	add_subdirectory ("glfw")
+    find_package(Freetype REQUIRED)
 endif()
-- 
2.56.0

